### PR TITLE
Fix convolution_depthwise3x3_winograd for multichannel output

### LINKED ISSUE #60176: Conv2D returns random results on arm64 with torch.no_grad()
## 🐛 Bug

Results of a convolution operation are different when used with and without torch.no_grad()

## To Reproduce

```bash
pip install --pre torch -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html
python 
```

```python3
import torch
import torch.nn as nn
patches = torch.rand(2,1,32,32)
m = nn.Conv2d(1, 32, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=True)
with torch.no_grad():
    output = m(patches)
print (output[0,0:10,0,0])
output2 = m(patches)
print (output2[0,0:10,0,0])
```

Prints:

```
tensor([-0.2287, -0.0931,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,
         0.0000,  0.0000])
tensor([-0.2287,  0.2208, -0.4023, -0.3761,  0.0098, -0.8294,  0.3598, -0.1206,
        -0.2089,  0.0942], grad_fn=<SelectBackward>)
```
## Expected behavior

Outputs should be the same, with grad and without it. 

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```
Collecting environment information...
PyTorch version: 1.10.0.dev20210616
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 11.4 (arm64)
GCC version: Could not collect
Clang version: 12.0.5 (clang-1205.0.22.9)
CMake version: version 3.20.3
Libc version: N/A

Python version: 3.9 (64-bit runtime)
Python platform: macOS-11.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.3
[pip3] torch==1.10.0.dev20210616
[conda] numpy                     1.20.3           py39h1f3b974_1    conda-forge
[conda] torch                     1.10.0.dev20210616          pypi_0    pypi

## Additional context

Possibly related issues: https://github.com/pytorch/pytorch/issues/59439, 

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @malfet

### PR DESCRIPTION
**Pull request by malfet**
Before this change it was implemented with the assumption, that number of groups, input  and output channels are the same, which is not always the case
Extend the implementation to support any number of output channels as long as number of groups equals to the number of input channels (i.e. kernel.size(1) == 1)

Fixes #60176

### CODE REVIEW

**Comment by albanD** [comment]
Ho, you mean that the check `https://github.com/pytorch/pytorch/blob/c50c77b444812d6f28bb7a9071ca0eb643ab6f78/aten/src/ATen/native/Convolution.cpp#L158` actually doesn't imply that the number of output channels is 1 but `groups`?

File: aten/src/ATen/native/cpu/DepthwiseConvKernel.cpp [hunk 1]
```diff
@@ -282,12 +284,13 @@ Tensor _convolution_depthwise3x3_winograd(
                       bias_potentially_undefined :
                       at::zeros({kernel_sizes[0]}, input.options());
 
-  at::parallel_for(0, args.batch * groups, 0, [&](int64_t start, int64_t end) {
+  at::parallel_for(0, args.batch * args.out_channels, 0, [&](int64_t start, int64_t end) {
```
[the remark this answers is not in the document]
Comment (malfet) [reply aten/src/ATen/native/cpu/DepthwiseConvKernel.cpp:287]: No, I think we a guaranteed that `groups == args.in_channels`, i.e. 
```
% python -c "import torch.nn as nn; print(nn.Conv2d(2, 32, (3, 3), groups=2).weight.shape)"
torch.Size([32, 1, 3, 3])
% python -c "import torch.nn as nn; print(nn.Conv2d(1, 32, (3, 3), groups=1).weight.shape)"
torch.Size([32, 1, 3, 3])
```

**Comment by malfet** [comment]
> Ho, you mean that the check
> 
> https://github.com/pytorch/pytorch/blob/c50c77b444812d6f28bb7a9071ca0eb643ab6f78/aten/src/ATen/native/Convolution.cpp#L158
> 
> actually doesn't imply that the number of output channels is 1 but `groups`?



No, it doesn't, see my comment above. `weight.channels == 1` means that there are 1<->1 mapping between input layer and output layer

File: aten/src/ATen/native/cpu/DepthwiseConvKernel.cpp [hunk 2]
```diff
@@ -22,6 +22,7 @@ struct Arguments final {
   // Output layer dimensions
   int64_t out_rows;
   int64_t out_cols;
+  int64_t out_channels;
```
Comment (albanD) [inline aten/src/ATen/native/cpu/DepthwiseConvKernel.cpp:25]: Why add it here as you don't actually pass it to the impl and only use it on the top level function?

File: aten/src/ATen/native/cpu/DepthwiseConvKernel.cpp [hunk 3]
```diff
@@ -282,12 +284,13 @@ Tensor _convolution_depthwise3x3_winograd(
                       bias_potentially_undefined :
                       at::zeros({kernel_sizes[0]}, input.options());
 
-  at::parallel_for(0, args.batch * groups, 0, [&](int64_t start, int64_t end) {
+  at::parallel_for(0, args.batch * args.out_channels, 0, [&](int64_t start, int64_t end) {
     for (int64_t k = start; k < end; ++k) {
-      const int64_t g = k % groups;
+      const int64_t g = k % args.out_channels;
+      const int64_t i = g % groups;
```
Comment (albanD) [inline aten/src/ATen/native/cpu/DepthwiseConvKernel.cpp:290]: From writing it on paper, this is what I get:
```suggestion
      const int64_t i = k / (args.out_channels / groups);
```
Namely, how many groups of outputs did you go through. That will correspond to the index after linearizing the batch size.

I'm really bad at modulo computation but the two are not equivalent right?

**Review by albanD — COMMENTED** [review]
Shall we add a test for this case to make sure the fix works?

File: aten/src/ATen/native/cpu/DepthwiseConvKernel.cpp (hunk 2 above)
Comment (malfet) [reply aten/src/ATen/native/cpu/DepthwiseConvKernel.cpp:25]: No particular reason, just keeping the pattern, can make it a local var if you prefer.

File: aten/src/ATen/native/cpu/DepthwiseConvKernel.cpp (hunk 3 above)
Comment (malfet) [reply aten/src/ATen/native/cpu/DepthwiseConvKernel.cpp:290]: You are right, two are not equivalent, but I think both formulas are wrong. Take `nn.Conv2d(1, 32, (3, 3))` for example. groups for such convolution would be 1, out_channels 32, but i should always be 0 if batch == 1, which is the case for my formula, but not for yours.

Comment (albanD) [reply aten/src/ATen/native/cpu/DepthwiseConvKernel.cpp:290]: Isn't it? k is in range `[0, out_channels[` for batch==1. So `k / out_channels` will always be 0

File: aten/src/ATen/native/cpu/DepthwiseConvKernel.cpp (hunk 2 above)
Comment (albanD) [reply aten/src/ATen/native/cpu/DepthwiseConvKernel.cpp:25]: It just creates a larger struct and pass it around for no reason.
But I agree it makes the code easier to read and it should be pretty insignificant impact so I'm happy to keep it as-is.

**Comment by malfet** [comment]
Addressed review feedback and added test

File: aten/src/ATen/native/cpu/DepthwiseConvKernel.cpp (hunk 3 above)
Comment (malfet) [reply aten/src/ATen/native/cpu/DepthwiseConvKernel.cpp]: You are right, it is a correct formula (using module would pick interleave input channels in the wrong order)

**Review by albanD — APPROVED** [review]
Thanks for the test.
LGTM

**Comment by malfet** [comment]
Fixed indent

**Pull request closed without merging**